projects/canopy/src/lib/utils/dynamic-style.service.ts

```diff
@@ -0,0 +1,70 @@
+import { Injectable } from '@angular/core';
+
+import { BreakpointValues } from '../shared/breakpoints.interface';
+
+@Injectable({
+  providedIn: 'root',
+})
+/**
+ * Dynamically add styles to a <style> tag in <head>
+ */
+export class DyanmicStyleService {
+  styleTag = null;
+  selectors = [];
+  mediaQueries = {};
+
+  private createStyleTag(): void {
+    this.styleTag = document.createElement('style');
+    this.styleTag.type = 'text/css';
+    document.getElementsByTagName('head')[0].appendChild(this.styleTag);
+  }
+
+  private addMediaQuery(breakpoint: BreakpointValues): void {
+    if (!this.styleTag) {
+      this.createStyleTag();
+    }
+    if (this.mediaQueries[breakpoint]) {
+      return;
+    }
+    this.mediaQueries[breakpoint] = [];
+    this.styleTag.innerHTML += `@media(min-width:${breakpoint}){}`;
+  }
+
+  addStyle(selector: string, style: string): void {
+    if (!this.styleTag) {
+      this.createStyleTag();
+    }
+    if (this.selectors.includes(selector)) {
+      return;
+    }
+    this.selectors.push(selector);
+    const styleString = `.${selector}{${style}!important}`;
+    this.styleTag.innerHTML += styleString;
+  }
+
+  addStyleToMediaQuery(
```

Comment:
I wondered about the two methods, not really focusing in implementation, but could one method be more intuitive? [Terminology wise](https://css-tricks.com/css-ruleset-terminology/) I wonder if it's a `rule` or `ruleset`

```ts
addRule(selector: string, declaration: string, breakpoint?: BreakpointValues)
```

Reply:
If you do manage to cut it down to one public method, you might be able to trim this repetition down to only be in that method...
```ts
if (!this.styleTag) {
  this.createStyleTag();
}
```

Reply:
Both methods are now called `addRules` and `addRulesToMediaQuery`. I didn't manage to combine the methods, as each one does a fair amount of different logic, that I didn't want to nest into a lot of `if` statements. To me it feels easier to read this way, but I'd be open to discuss.